### ISSUE TITLE
Oneshot Service stays in "activating" state breaking timers

### ISSUE THREAD

**Issue by danielkrajnik** [systemd/systemd#28698]
### systemd version the issue has been seen with

systemd 254 (254-1-arch)

### Used distribution

EndeavourOS

### Linux kernel version used

6.4.7-arch1-3

### CPU architectures issue was seen on

x86_64

### Expected behaviour you didn't see

timer with `OnActiveUnitSec=` repeat

### Unexpected behaviour you saw

timer with `OnActiveUnitSec=` stops repeating 

### Steps to reproduce the problem

have a `oneshot` unit that exits incorrectly (ending up in `Active: activating (start)` state).
timers stop repeating (I assume they keep waiting for the unit to finish correctly)

This caused a lot of missed backups. I don't know what causes the unit to end up in this state, but backups WORK FINE. Systemd in this case introduces a silent, implicit behaviour that user shouldn't be burdened with.

This greatly complicated usage of systemd timers. You shouldn't have to be an expert in systemd states to perform simple tasks.

If timer is set OnCalendar it should continue working on Calendar even if the unit is in `Active: activating (start)` state.

**Comment by poettering** [comment]
plese provide logs/systemctl status output when this happened. Otherwise this is not actionable.

Generally, if the service remains in "activating" state we'll not trigger another run.

The question is why it is stuck in "activating" state. For some reason your service seems to never complete. Logs + systemctl status output will let you know why.

**Comment by danielkrajnik** [comment]
log:

```
...
Aug 07 20:54:42 host restic[1400]: error: open /path1/xrdp/cert.pem: permission denied
Aug 07 20:54:42 host restic[1400]: error: open /path1/xrdp/key.pem: permission denied
Aug 07 20:54:42 host restic[1400]: error: open /path1/xrdp/rsakeys.ini: permission denied
Aug 07 20:57:02 host restic[1400]: Files:          29 new,   150 changed, 583372 unmodified
Aug 07 20:57:02 host restic[1400]: Dirs:            6 new, 51016 changed,   351 unmodified
Aug 07 20:57:02 host restic[1400]: Added to the repository: 97.728 MiB (97.855 MiB stored)
Aug 07 20:57:02 host restic[1400]: processed 583551 files, 118.895 GiB in 2:29
Aug 07 20:57:02 host restic[1400]: snapshot 210d97b5 saved
Aug 07 20:57:02 host restic[1400]: Warning: at least one source file could not be read
Aug 07 20:57:13 host restic[5517]: Applying Policy: keep 7 daily, 3 weekly, 6 monthly, 10 yearly snapshots
Aug 07 20:57:13 host restic[5517]: keep 4 snapshots:
Aug 07 20:57:13 host restic[5517]: ID        Time                 Host        Tags           Reasons           Paths
Aug 07 20:57:13 host restic[5517]: ---------------------------------------------------------------------------------------
Aug 07 20:57:13 host restic[5517]: 785ac673  2023-08-04 22:26:43  host3       systemd.timer  daily snapshot    /path1
Aug 07 20:57:13 host restic[5517]:                                                           weekly snapshot   /path2
Aug 07 20:57:13 host restic[5517]:                                                           monthly snapshot  /path3
Aug 07 20:57:13 host restic[5517]:                                                           yearly snapshot
Aug 07 20:57:13 host restic[5517]: fba27b0a  2023-08-05 15:41:53  host3       systemd.timer  daily snapshot    /path1
Aug 07 20:57:13 host restic[5517]:                                                                             /path2
Aug 07 20:57:13 host restic[5517]:                                                                             /path3
Aug 07 20:57:13 host restic[5517]: dab4ebb3  2023-08-06 23:00:27  host3       systemd.timer  daily snapshot    /path1
Aug 07 20:57:13 host restic[5517]:                                                           weekly snapshot   /path2
Aug 07 20:57:13 host restic[5517]:                                                                             /path3
Aug 07 20:57:13 host restic[5517]: 210d97b5  2023-08-07 20:54:28  host3       systemd.timer  daily snapshot    /path1
Aug 07 20:57:13 host restic[5517]:                                                           weekly snapshot   /path2
Aug 07 20:57:13 host restic[5517]:                                                           monthly snapshot  /path3
Aug 07 20:57:13 host restic[5517]:                                                           yearly snapshot
Aug 07 20:57:13 host restic[5517]: ---------------------------------------------------------------------------------------
Aug 07 20:57:13 host restic[5517]: 4 snapshots
Aug 07 20:57:13 host restic[5517]: remove 1 snapshots:
Aug 07 20:57:13 host restic[5517]: ID        Time                 Host        Tags           Paths
Aug 07 20:57:13 host restic[5517]: ---------------------------------------------------------------------
Aug 07 20:57:13 host restic[5517]: 2e6fb375  2023-08-07 20:10:00  host3       systemd.timer  /path1
Aug 07 20:57:13 host restic[5517]:                                                           /path2
Aug 07 20:57:13 host restic[5517]:                                                           /path3
Aug 07 20:57:13 host restic[5517]: ---------------------------------------------------------------------
Aug 07 20:57:13 host restic[5517]: 1 snapshots
Aug 07 20:57:13 host restic[5517]: [0:00] 100.00%  1 / 1 files deleted
Aug 07 20:57:13 host restic[5517]: 1 snapshots have been removed, running prune
Aug 07 20:57:13 host restic[5517]: loading indexes...
Aug 07 20:57:16 host restic[5517]: loading all snapshots...
Aug 07 20:57:16 host restic[5517]: finding data that is still in use for 29 snapshots
Aug 07 20:57:49 host restic[5517]: [0:32] 100.00%  29 / 29 snapshots
Aug 07 20:57:49 host restic[5517]: searching used packs...
Aug 07 20:57:50 host restic[5517]: collecting packs for deletion and repacking
Aug 07 20:58:18 host restic[5517]: [0:28] 100.00%  40088 / 40088 packs processed
Aug 07 20:58:18 host restic[5517]: to repack:         10828 blobs / 121.741 MiB
Aug 07 20:58:18 host restic[5517]: this removes:      10706 blobs / 113.079 MiB
Aug 07 20:58:18 host restic[5517]: to delete:         37307 blobs / 294.069 MiB
Aug 07 20:58:18 host restic[5517]: total prune:       48013 blobs / 407.147 MiB
Aug 07 20:58:18 host restic[5517]: remaining:       1841433 blobs / 232.174 GiB
Aug 07 20:58:18 host restic[5517]: unused size after prune: 11.604 GiB (5.00% of remaining size)
Aug 07 20:58:18 host restic[5517]: deleting unreferenced packs
Aug 07 20:58:19 host restic[5517]: [0:00] 100.00%  1 / 1 files deleted
Aug 07 20:58:19 host restic[5517]: repacking packs
Aug 07 20:58:59 host restic[5517]: [0:40] 100.00%  8 / 8 packs repacked
Aug 07 20:58:59 host restic[5517]: rebuilding index
Aug 07 21:02:31 host restic[5517]: [3:31] 100.00%  40065 / 40065 packs processed
Aug 07 21:02:31 host restic[5517]: deleting obsolete index files
Aug 07 21:02:36 host restic[5517]: [0:05] 100.00%  38 / 38 files deleted
Aug 07 21:02:36 host restic[5517]: removing 25 old packs
Aug 07 21:02:40 host restic[5517]: [0:03] 100.00%  25 / 25 files deleted
Aug 07 21:02:40 host restic[5517]: done
```

> Generally, if the service remains in "activating" state we'll not trigger another run.
cron would trigger it. Would it be possible to add such an option? It would make it much simpler.

> The question is why it is stuck in "activating" state.

I don't know, but it will be really complex to debug:

````
[Unit]
After = network.target

[Service]
Type=oneshot
ExecStart=-restic backup /path2 /path1 /path3 --one-file-system --exclude /home/user/.cache  --repo rclone:gdrive:backups --tag systemd.timer
ExecStart=systemctl stop --user mount-restic-backup.service
ExecStart=restic unlock --repo rclone:gdrive:backups
ExecStart=restic forget --tag systemd.timer --keep-daily 7 --keep-weekly 3 --keep-monthly 6 --keep-yearly 10 --prune --repo rclone:gdrive:backups 
ExecStart=systemctl start --user mount-restic-backup.service

[Install]
WantedBy=default.target
```

if it was possible to just say "run irrespective of previous run" that would be really really helpful

**Comment by poettering** [comment]
Your last ExecStart= line is synchronous, i.e. systemctl start will wait for the specified service to start, and thus block the whole unit on completion of it. Is that your intention? If not add --no-block to the systemctl cmdline.

**Comment by dtardon** [comment]
> if it was possible to just say "run irrespective of previous run" that would be really really helpful

#16886

**Comment by poettering** [comment]
> Thank you, I wasn't aware of the `--no-block` option - trying it now. If it wasn't a systemctl, but some other program, could you precede it with a hyphen instead? `ExecStart=-systemctl start --user mount-restic-backup.service`? Would that have the same effect?

No. The "-" thing is about ignoring exit codes and considered even failure exit codes of the specified program as success.

The --no-block switch just makes systemctl operate async, i.e. it will enqueue the job and exit, rather than also wait for the job to complete. 
 
> I think that it would still be useful to tell systemd to keep going with the timer even if some part of the service isn't in correct state. You wouldn't have to worry about some unexpected bugs in the future. Sometimes you just want to let it run irrespective of its side effects.

I am not convinced that's a good idea, since this would pile up services if the repetition intervals are shorter than the the scripts run.

**Comment by poettering** [comment]
Anyway, closing, since the issue at hand is apparently addressed.

**Comment by danielkrajnik** [comment]
Sorry, it keeps happening. Service gets stuck in activating state and backups stop.

This is too annoying to keep track of. I strongly recommend anyone who needs reliable a (so everyone basically) to use cron instead.

Systemd timers became too complicated for certain use cases. If you need to update ssl certificate once very year that's fine, but it completely FAILS with more complex use cases on desktop.

It is suprising though because backup is a very common use case and it's disappointing to see systemd timers fail to accommodate modern solutions (rclone with restic).

**Comment by danielkrajnik** [comment]
@TriMoon hmm you may be right.I first started writing cron commands inline (directly in `crontab -e`), but a substantial number of them turned out to be complicated enough to grant writing them in separate shell script files.

However, I still remain convinced that none of these scripts require dependency control and process babysitting introduced by systemd's timers. These features are great for init, but they caused me a lot of issues some of which I've never resolved. 

## tldr; cron still good
